Fix corrupted negative numbers when exporting CSV

Jamim opened the pull request:
Hello,

I hope you'll find this PR useful.
It fixes #18177.

These changes also:
 - Make `formatRow` function keep cells empty for `null` and `undefined` values
 - Remove usage of the deprecated `isNullOrUndefined`

Best regards!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=18179) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=18179) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=18179) it.</sub>

torkelo commented:
Might be safer to check if the actual value is a number, instead of column type

Jamim commented:
Hello @torkelo,

Thank you for the feedback, but I believe that checking of an actual value type instead of a column type is not safer.
At least, for the `date` column type actual type of the value is `number` and it should still be formatted as the `date`.

![DevTools](https://user-images.githubusercontent.com/5607572/61872018-f5260280-aeea-11e9-82ef-376770457990.png)

To be honest, I don't see any benefits from checking of the actual value type since, as far as I can see, number formatters just keep non-numeric values "as is".

Jamim committed:
Fix corrupted negative numbers when exporting CSV

Fixes #18177

These changes also:

 - Make formatRow function keep cells empty for null and undefined values

 - Remove usage of the deprecated isNullOrUndefined

torkelo commented on public/app/core/specs/file_export.test.ts:
```diff
@@ -53,10 +53,10 @@ describe('file_export', () => {
-        '"1500026400";null;null\r\n' +
-        '"1500026500";null;15\r\n' +
-        '"1500026600";6;null';
+        '"1500026300";;13\r\n' +
```
this looks wrong, no longer outputting null

Jamim replied:
I mentioned this change in the description and in the commit message.
And I leaved [a corresponding comment](https://github.com/grafana/grafana/pull/18179/files#diff-9d9218dee66e94b7d96fa5d7cb19ab83R57) in the `formatRow` as well.

The idea here is to keep the resulting CSV readable and exclude meaningless values.

Jamim replied:
Currently, without these changes, null values are rendered as `"'-"` which is a garbage.

torkelo replied:
They should be rendered as null so it’s clear it’s null

Jamim replied:
> They should be rendered as null so it’s clear it’s null

This is not how the export works now because all values are currently formatted regardless of their types.

Anyway, is there any good reason to render `null` values as `null` strings in CSV instead of keeping corresponding cells empty?
CSV in this case looks strange or even broken for any end user who is not a JS/TS programmer.

Jamim replied:
Let me repeat my question.
> is there any good reason to render `null` values as `null` strings in CSV instead of keeping corresponding cells empty?

DanCech commented:
If I'm understanding this correctly, the problem is that when exporting data from the table panel we're exporting the formatted values which are all strings.  Because they're all strings, the csv exporter treats all values as strings and escapes them.

The change in `table/renderer.ts` skips formatting values for "number" columns, which would mean that the values would be returned as-is in the csv export (assuming they are in fact returned as numbers and not text values).  This seems like it would be a reasonable solution to the problem, and it may actually make sense to use something like the lodash `toNumber` function to ensure that the values are in fact numeric.

With that modification the csv export code itself wouldn't need to be changed, as is already handles numbers correctly.

ryantxu commented:
I am agnostic on how this gets fixed for file_export.ts

After we have DataFrame transforms implemented (series_to_rows, etc) then this should be replaced with [toCSV( data )](https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/utils/csv.ts#L318).

Jamim commented:
Hello @ryantxu,

> I am agnostic on how this gets fixed for file_export.ts

Sorry, but I don't understand what you mean.

ryantxu commented:
> Sorry, but I don't understand what you mean.

Just that `file_export.ts` should be thrown away very soon and will soon use [toCSV( data )](https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/utils/csv.ts#L318).  

In that path, we know what "type" a column is so we can better support the behavior we think makes sense.
